### PR TITLE
feat: add appeal config to policy

### PR DESCRIPTION
**Pull request by haveiss**
closes https://github.com/odpf/guardian/issues/226

the approach that we discussed is:
1. keep appeal config in provider
2. introduce `allow_permanent_access` and `allow_active_access_extension_in` in policy.appeal_config
3. on the appeal creation business logic, policy.appeal_config will be prioritized over provider.appeal_config
4. during this version/release is live, users need to enter the appeal config in their policies.
5. and then we will create a next release containing provider.appeal_config deprecation/removal

related proton PR: https://github.com/odpf/proton/pull/205

### CODE REVIEW

File: api/handler/v1beta1/provider.go [hunk 1]
```diff
@@ -147,35 +146,3 @@ func (s *GRPCServer) ListRoles(ctx context.Context, req *guardianv1beta1.ListRol
...
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "failed to parse appeal proto %q: %v", g.ID, err)
-		}
-		grantsProto = append(grantsProto, grantProto)
-	}
-
-	return &guardianv1beta1.ImportAccessResponse{
-		Grants: grantsProto,
-	}, nil
-}
```
Comment (haveiss) [inline api/handler/v1beta1/provider.go:181]: waiting for https://github.com/odpf/proton/pull/197 to be updated and merged

File: core/appeal/service.go [hunk 2]
```diff
@@ -232,15 +240,8 @@ func (s *Service) Create(ctx context.Context, appeals []*domain.Appeal, opts ...
...
-		if isAdditionalAppealCreation && appeal.PolicyID != "" && appeal.PolicyVersion != 0 {
-			policy = policies[appeal.PolicyID][appeal.PolicyVersion]
-		} else {
-			var err error
-			policy, err = getPolicy(appeal, provider, policies)
-			if err != nil {
-				return fmt.Errorf("retrieving policy: %w", err)
-			}
+		if err2 != nil {
+			return fmt.Errorf("retrieving policy: %w", err2)
```
Comment (haveiss) [inline core/appeal/service.go:244]: add more test scenario

File: core/provider/service.go [hunk 3]
```diff
@@ -267,7 +267,13 @@ func (s *Service) ValidateAppeal(ctx context.Context, a *domain.Appeal, p *domai
...
 	}
 
-	if p.Config.Appeal != nil && !p.Config.Appeal.AllowPermanentAccess {
+	// Default to use provider config if policy config is not set
+	AllowPermanentAccess := p.Config.Appeal.AllowPermanentAccess
+	if policy != nil && policy.AppealConfig != nil {
+		AllowPermanentAccess = policy.AppealConfig.AllowPermanentAccess
+	}
+
+	if p.Config.Appeal != nil && !AllowPermanentAccess {
```
Comment (haveiss) [inline core/provider/service.go:276]: add test scenario

File: core/appeal/service.go [hunk 4]
```diff
@@ -218,11 +218,19 @@ func (s *Service) Create(ctx context.Context, appeals []*domain.Appeal, opts ...
 			return fmt.Errorf("retrieving provider: %w", err)
 		}
 
-		if err := s.checkExtensionEligibility(appeal, provider, activeGrants); err != nil {
+		var err2 error
+		var policy *domain.Policy
+		if isAdditionalAppealCreation && appeal.PolicyID != "" && appeal.PolicyVersion != 0 {
+			policy = policies[appeal.PolicyID][appeal.PolicyVersion]
+		} else {
+			policy, err2 = getPolicy(appeal, provider, policies)
```
Comment (mabdh) [inline core/appeal/service.go:226]: why not return error here instead if `err2 != nil`

File: core/appeal/service.go [hunk 5]
```diff
@@ -1003,18 +1004,26 @@ func (s *Service) CreateAccess(ctx context.Context, a *domain.Appeal, opts ...Cr
...
+			AllowActiveAccessExtensionIn := p.Config.Appeal.AllowActiveAccessExtensionIn
+			if policy != nil &&
+				policy.AppealConfig != nil &&
+				policy.AppealConfig.AllowActiveAccessExtensionIn != "" {
+				AllowActiveAccessExtensionIn = policy.AppealConfig.AllowActiveAccessExtensionIn
+			}
+
+			if AllowActiveAccessExtensionIn == "" {
 				return ErrAppealFoundActiveGrant
 			}
```
Comment (mabdh) [inline core/appeal/service.go:1022]: I think we could also return error early if `AllowActiveAccessExtensionIn` in empty
```suggestion
			// Default to use provider config if policy config is not set
			AllowActiveAccessExtensionIn := p.Config.Appeal.AllowActiveAccessExtensionIn
			if AllowActiveAccessExtensionIn == "" {
				return ErrAppealFoundActiveGrant
			}
			
			if policy != nil &&
				policy.AppealConfig != nil &&
				policy.AppealConfig.AllowActiveAccessExtensionIn != "" {
				AllowActiveAccessExtensionIn = policy.AppealConfig.AllowActiveAccessExtensionIn
			}

```

File: core/policy/service.go [hunk 6]
```diff
@@ -366,3 +371,19 @@ func structToMap(item interface{}) (map[string]interface{}, error) {
 
 	return result, nil
 }
+
+func (s *Service) validateAppealConfig(cfg *domain.PolicyAppealConfig) error {
+	if cfg != nil && cfg.AllowActiveAccessExtensionIn != "" {
+		if err := validateDuration(cfg.AllowActiveAccessExtensionIn); err != nil {
+			return fmt.Errorf("invalid appeal extension policy: %v", err)
```
Comment (mabdh) [inline core/policy/service.go:378]: error wrapping with `%w` ?
```suggestion
			return fmt.Errorf("invalid appeal extension policy: %w", err)
```

File: core/policy/service.go [hunk 7]
```diff
@@ -366,3 +371,19 @@ func structToMap(item interface{}) (map[string]interface{}, error) {
...
+		}
+	}
+	return nil
+}
+
+func validateDuration(d string) error {
+	if _, err := time.ParseDuration(d); err != nil {
+		return fmt.Errorf("parsing duration: %v", err)
```
Comment (mabdh) [inline core/policy/service.go:386]: same with this, we can use `%w` for error wrapping

File: core/appeal/service.go (hunk 4 above)
Comment (haveiss) [reply core/appeal/service.go:226]: yea, initially i did that, but unfortunately breaking existing tests due to validation order,
so i keep the existing behavior by keeping validation order the same,
do you have any suggestion mas @mabdh ?

Comment (mabdh) [reply core/appeal/service.go:226]: Hmm that is fair enough. But seeing the flow, the validation is unnecessary if `policy` is nil right ? 
Should we also re-evaluate the testcase itself?

**Review by mabdh — APPROVED** [review]

**Pull request merged**